File: kakao.app/js/ked/view/colmns/funcol.js
var _k_ = {extend: function (c,p) {for (var k in p) { if (Object.hasOwn(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}, empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}, max: function () { var m = -Infinity; for (var a of arguments) { if (Array.isArray(a)) {m = _k_.max.apply(_k_.max,[m].concat(a))} else {var n = parseFloat(a); if(!isNaN(n)){m = n > m ? n : m}}}; return m }}

var funcol

import kxk from "../../../kxk.js"
let post = kxk.post
let slash = kxk.slash

import color from "../../theme/color.js"
import theme from "../../theme/theme.js"

import view from "../base/view.js"
import knob from "../base/knob.js"

import funtree from "./funtree.js"


funcol = (function ()
{
    _k_.extend(funcol, view)
    function funcol (screen, editor, features)
    {
        this.editor = editor
    
        this["onKey"] = this["onKey"].bind(this)
        this["onWheel"] = this["onWheel"].bind(this)
        this["onMouse"] = this["onMouse"].bind(this)
        this["onFuncolToggle"] = this["onFuncolToggle"].bind(this)
        this["onFuncolResize"] = this["onFuncolResize"].bind(this)
        this["onContext"] = this["onContext"].bind(this)
        this["onSessionMerge"] = this["onSessionMerge"].bind(this)
        funcol.__super__.constructor.call(this,screen,'funcol',features)
        this.isVisible = false
        this.active = false
        this.pointerType = 'pointer'
        this.knob = new knob(screen,`${this.name}_knob`)
        this.funtree = new funtree(this.editor,`${this.name}_funtree`,['scrllr'])
        this.knob.frameSide = 'left'
        this.funtree.setColor('bg',theme.funtree.bg)
        this.funtree.setColor('empty',this.funtree.color.bg)
        this.funtree.setColor('cursor_main',this.funtree.color.bg)
        this.funtree.setColor('cursor_empty',this.funtree.color.bg)
        this.funtree.scroll.setColor('bg',this.funtree.color.bg)
        this.knob.setColor('bg',this.funtree.color.bg)
        post.on('funcol.resize',this.onFuncolResize)
        post.on('funcol.toggle',this.onFuncolToggle)
        post.on('session.merge',this.onSessionMerge)
    }

    funcol.prototype["onSessionMerge"] = function (recent)
    {
        if (_k_.empty(recent.funcol))
        {
            return
        }
        if (recent.funcol.active)
        {
            this.active = true
            this.show()
        }
        return ked_session.set('funcol',recent.funcol)
    }

    funcol.prototype["layout"] = function (x, y, w, h)
    {
        this.funtree.layout(x,y,w,h)
        this.knob.layout(x,y,1,h)
        return funcol.__super__.layout.call(this,x,y,w,h)
    }

    funcol.prototype["draw"] = function ()
    {
        if (this.hidden() || this.collapsed() || !this.active)
        {
            return
        }
        this.cells.fill_rect(0,1,-1,-1,' ',null,theme.funtree.bg)
        this.cells.fill_rect(0,0,-1,0,' ',null,theme.gutter.bg)
        this.funtree.draw()
        this.knob.draw()
        return funcol.__super__.draw.call(this)
    }

    funcol.prototype["onContext"] = function (event)
    {
        if (!this.hover)
        {
            return
        }
        console.log(`funcol.onContext: ${this.hover}`,event)
    }

    funcol.prototype["onFuncolResize"] = function ()
    {
        return this.knob.doDrag = true
    }

    funcol.prototype["onFuncolToggle"] = function ()
    {
        var cols

        if (!(this.visible() && this.collapsed()))
        {
            this.toggle()
        }
        this.active = this.visible()
        ked_session.set('funcol▸active',this.active)
        cols = _k_.max(16,parseInt(this.cells.screen.cols / 6))
        return post.emit('view.size',this.name,'left',((this.hidden() ? -this.cells.cols : cols - this.cells.cols)))
    }

    funcol.prototype["onMouse"] = function (event)
    {
        var ret

        if (this.hidden() || this.collapsed() || !this.active)
        {
            return
        }
        ret = funcol.__super__.onMouse.call(this,event)
        if ((ret != null ? ret.redraw : undefined))
        {
            return ret
        }
        ret = this.knob.onMouse(event)
        if ((ret != null ? ret.redraw : undefined))
        {
            return ret
        }
        ret = this.funtree.onMouse(event)
        if ((ret != null ? ret.redraw : undefined))
        {
            return ret
        }
    }

    funcol.prototype["onWheel"] = function (event)
    {
        if (this.hidden() || this.collapsed() || !this.active)
        {
            return
        }
        return this.funtree.onWheel(event)
    }

    funcol.prototype["onKey"] = function (key, event)
    {
        if (!this.funtree.hasFocus())
        {
            return
        }
        return this.funtree.onKey(key,event)
    }

    return funcol
})()

export default funcol;

File: kakao.app/love/lua/view/colmns/funcol.lua
--[[
00000000  000   000  000   000   0000000   0000000   000    
000       000   000  0000  000  000       000   000  000    
000000    000   000  000 0 000  000       000   000  000    
000       000   000  000  0000  000       000   000  000    
000        0000000   000   000   0000000   0000000   0000000
--]]

funtree = require "view.colmns.funtree"


local funcol = class("funcol", view)
    


function funcol:init(editor, features) 
        self.editor = editor
        
        view.init(self, 'funcol', features)
        
        self.isVisible = false
        self.active = false
        
        self.pointerType = 'pointer'
        
        self.knob = knob("" .. tostring(self.name) .. "_knob")
        self.funtree = funtree(self.editor, "" .. tostring(self.name) .. "_funtree", array('scrllr'))
        
        self.knob.frameSide = 'left'
        
        self.funtree:setColor('bg', theme.funtree.bg)
        self.funtree:setColor('empty', self.funtree.color.bg)
        self.funtree:setColor('cursor_main', self.funtree.color.bg)
        self.funtree:setColor('cursor_empty', self.funtree.color.bg)
        self.funtree.scroll:setColor('bg', self.funtree.color.bg)
        
        self.knob:setColor('bg', self.funtree.color.bg)
        
        post:on('funcol.resize', self.onFuncolResize)
        post:on('funcol.toggle', self.onFuncolToggle)
        post:on('session.merge', self.onSessionMerge)
        return self
    end


function funcol:onSessionMerge(recent) 
        if empty(recent.funcol) then return end
        
        if recent.funcol.active then 
            self.active = true
            self:show()
        end
        
        return ked_session:set('funcol', recent.funcol)
    end

-- ███       ███████   ███   ███   ███████   ███   ███  █████████
-- ███      ███   ███   ███ ███   ███   ███  ███   ███     ███   
-- ███      █████████    █████    ███   ███  ███   ███     ███   
-- ███      ███   ███     ███     ███   ███  ███   ███     ███   
-- ███████  ███   ███     ███      ███████    ███████      ███   


function funcol:layout(x, y, w, h) 
        self.funtree:layout(x, y, w, h)
        self.knob:layout(x, y, 1, h)
        
        view.layout(self, x, y, w, h)
        return self
    end

-- ███████    ████████    ███████   ███   ███
-- ███   ███  ███   ███  ███   ███  ███ █ ███
-- ███   ███  ███████    █████████  █████████
-- ███   ███  ███   ███  ███   ███  ███   ███
-- ███████    ███   ███  ███   ███  ██     ██


function funcol:draw() 
        if ((self:hidden() or self:collapsed()) or not self.active) then return end
        
        self.cells:fill_rect(0, 1, -1, -1, ' ', nil, theme.funtree.bg)
        self.cells:fill_rect(0, 0, -1, 0, ' ', nil, theme.gutter.bg)
        
        self.funtree:draw()
        self.knob:draw()
        
        view.draw(self)
        return self
    end

--  ███████   ███████   ███   ███  █████████  ████████  ███   ███  █████████
-- ███       ███   ███  ████  ███     ███     ███        ███ ███      ███   
-- ███       ███   ███  ███ █ ███     ███     ███████     █████       ███   
-- ███       ███   ███  ███  ████     ███     ███        ███ ███      ███   
--  ███████   ███████   ███   ███     ███     ████████  ███   ███     ███   


function funcol:onContext(event) 
        if not self.hover then return end
        
        return print("funcol.onContext: " .. tostring(self.hover) .. "", event)
    end

-- █████████   ███████    ███████    ███████   ███      ████████
--    ███     ███   ███  ███        ███        ███      ███     
--    ███     ███   ███  ███  ████  ███  ████  ███      ███████ 
--    ███     ███   ███  ███   ███  ███   ███  ███      ███     
--    ███      ███████    ███████    ███████   ███████  ████████


function funcol:onFuncolResize() 
    self.knob.doDrag = true
    return self.knob.doDrag
    end

function funcol:onFuncolToggle() 
        if not (self:visible() and self:collapsed()) then self:toggle() end
        self.active = self:visible()
        ked_session:set('funcol▸active', self.active)
        local cols = max(16, int((self.cells.screen.cols / 6)))
        return -- post.emit 'view.size' @name 'left' (@hidden() ? -@cells.cols : cols-@cells.cols)
    end

-- 00     00   0000000   000   000   0000000  00000000  
-- 000   000  000   000  000   000  000       000       
-- 000000000  000   000  000   000  0000000   0000000   
-- 000 0 000  000   000  000   000       000  000       
-- 000   000   0000000    0000000   0000000   00000000  


function funcol:onMouse(event) 
        if ((self:hidden() or self:collapsed()) or not self.active) then return end
        
        if view.onMouse(self, event) then return true end
        if self.knob:onMouse(event) then return true end
        if self.funtree:onMouse(event) then 
    return true
        end
    end

-- 000   000  000   000  00000000  00000000  000      
-- 000 0 000  000   000  000       000       000      
-- 000000000  000000000  0000000   0000000   000      
-- 000   000  000   000  000       000       000      
-- 00     00  000   000  00000000  00000000  0000000  


function funcol:onWheel(event) 
        if ((self:hidden() or self:collapsed()) or not self.active) then return end
        return self.funtree:onWheel(event)
    end

-- 000   000  00000000  000   000  
-- 000  000   000        000 000   
-- 0000000    0000000     00000    
-- 000  000   000          000     
-- 000   000  00000000     000     


function funcol:onKey(key, event) 
        if not self.funtree:hasFocus() then return end
        
        return self.funtree:onKey(key, event)
    end

return funcol